crates/bevy_pbr/src/material.rs

```diff
@@ -1,321 +1,312 @@
-use crate::{AlphaMode, PbrPipeline, StandardMaterialFlags};
+use crate::{
+    AlphaMode, DrawMesh, MeshPipeline, MeshPipelineKey, MeshUniform, SetMeshBindGroup,
+    SetMeshViewBindGroup,
+};
 use bevy_app::{App, Plugin};
-use bevy_asset::{AddAsset, Handle, HandleUntyped};
-use bevy_ecs::system::{lifetimeless::SRes, SystemParamItem};
-use bevy_math::Vec4;
-use bevy_reflect::TypeUuid;
+use bevy_asset::{AddAsset, Asset, AssetServer, Handle};
+use bevy_core_pipeline::{AlphaMask3d, Opaque3d, Transparent3d};
+use bevy_ecs::{
+    entity::Entity,
+    prelude::World,
+    system::{
+        lifetimeless::{Read, SQuery, SRes},
+        Query, Res, ResMut, SystemParamItem,
+    },
+    world::FromWorld,
+};
 use bevy_render::{
-    color::Color,
-    render_asset::{PrepareAssetError, RenderAsset, RenderAssetPlugin, RenderAssets},
+    mesh::Mesh,
+    render_asset::{RenderAsset, RenderAssetPlugin, RenderAssets},
+    render_component::ExtractComponentPlugin,
+    render_phase::{
+        AddRenderCommand, DrawFunctions, EntityRenderCommand, RenderCommandResult, RenderPhase,
+        SetItemPipeline, TrackedRenderPass,
+    },
     render_resource::{
-        std140::{AsStd140, Std140},
-        BindGroup, Buffer, BufferInitDescriptor, BufferUsages,
+        BindGroup, BindGroupLayout, RenderPipelineCache, RenderPipelineDescriptor, Shader,
+        SpecializedPipeline, SpecializedPipelines,
     },
     renderer::RenderDevice,
-    texture::Image,
+    view::{ExtractedView, Msaa, VisibleEntities},
+    RenderApp, RenderStage,
 };
-use wgpu::{BindGroupDescriptor, BindGroupEntry, BindingResource};
+use std::hash::Hash;
+use std::marker::PhantomData;
+
+pub trait Material: Asset + RenderAsset {
+    fn bind_group(render_asset: &<Self as RenderAsset>::PreparedAsset) -> &BindGroup;
+    fn bind_group_layout(render_device: &RenderDevice) -> BindGroupLayout;
+    #[allow(unused_variables)]
+    fn vertex_shader(asset_server: &AssetServer) -> Option<Handle<Shader>> {
+        None
+    }
+    #[allow(unused_variables)]
+    fn fragment_shader(asset_server: &AssetServer) -> Option<Handle<Shader>> {
+        None
+    }
+    #[allow(unused_variables)]
+    fn alpha_mode(render_asset: &<Self as RenderAsset>::PreparedAsset) -> AlphaMode {
+        AlphaMode::Opaque
+    }
+}
+
+impl<M: Material> SpecializedMaterial for M {
+    type Key = ();
+
```

Comment:
maybe add a `type AdditionalRenderCommands` as well?
That would enable the trait to be used in more situations where for example more bind groups are used or a second vertex buffer is used for instancing.

Reply:
I think we want the trait to be scoped to a single bind group / material because ultimately we'll want to add opt-in instancing for materials. Adding more flexibility would remove our ability to generate batches of things to instance (or at the very least, make it prohibitively complicated)

I also don't think more type complexity here is a good look. We need to be going in the opposite direction to optimize for the common case (with things like Bind Group derives)

I think theres room for more than one high level abstraction. I think its worth trying to come up with something less constrained (but probably with more boilerplate). If we can ultimately unify everything under one abstraction ... awesome ... but I doubt we can do that in a way that satisfies all of our requirements (flexibility / terseness / efficiency).

Reply:
Resolving this for now because I think expanding the scope is risky (especially in the short term).